src/main/resources/templates/db_scripts/itemEffectiveLocation.sql

```diff
@@ -0,0 +1,86 @@
+
+-- Updates item effective location property in case holding's effective location has changed.
+-- Only items that do not have permanent or temp location set will be updated.
+CREATE OR REPLACE FUNCTION ${myuniversity}_${mymodule}.update_effective_location_on_holding_update() RETURNS trigger
+AS $$
+  DECLARE
+    old_effective_location_id jsonb;
+    new_effective_location_id jsonb;
+  BEGIN
+    old_effective_location_id = coalesce(OLD.jsonb->'temporaryLocationId', OLD.jsonb->'permanentLocationId');
+    new_effective_location_id = coalesce(NEW.jsonb->'temporaryLocationId', NEW.jsonb->'permanentLocationId');
+    -- null-safe comparison
+    if (new_effective_location_id IS NOT DISTINCT FROM old_effective_location_id) THEN
+      RETURN NEW;
+    end if;
+    if (new_effective_location_id IS NULL) then
+      -- empty string signals the item trigger to remove the effective location without running a holdings query
+      new_effective_location_id = to_jsonb(''::text);
+    end if;
+    UPDATE ${myuniversity}_${mymodule}.item
+      SET jsonb = jsonb_set(jsonb, '{effectiveLocationId}', new_effective_location_id)
+      WHERE holdingsRecordId = new.id
+            AND permanentLocationId IS NULL
+            AND temporaryLocationId IS NULL;
+    -- if item's either temp or perm location is not null then we don't need to update it's effective location
+    -- as item's location has more priority than holding's one.
+    RETURN NEW;
+  END;
+  $$ LANGUAGE 'plpgsql';
+
+-- Set Item effective location in case insert or update of temporary/permanentLocationId or holdingsRecordId properties
+CREATE OR REPLACE FUNCTION ${myuniversity}_${mymodule}.update_effective_location_on_item_update() RETURNS trigger
+AS $$
+  declare
+    effective_location_id jsonb;
+  begin
+    effective_location_id = coalesce(NEW.jsonb->'temporaryLocationId', NEW.jsonb->'permanentLocationId');
+    if (effective_location_id IS NOT NULL) then
+      NEW.jsonb = jsonb_set(NEW.jsonb, '{effectiveLocationId}', effective_location_id);
+      return NEW;
+    end if;
+    effective_location_id = NEW.jsonb->'effectiveLocationId';
+    if (effective_location_id IS NOT NULL) then
+      -- Field exists so we know it was set by the holdings trigger
+      -- because RMB always removes this read-only field.
+      if (effective_location_id ? '') then
+        -- Holdings trigger uses empty string to signal removal, so there is no need to run a holdings query.
+        -- "effectiveLocationId" must be a valid UUID or removed at all, null or empty string is not allowed.
+        NEW.jsonb = NEW.jsonb - 'effectiveLocationId';
+      end if;
+      return NEW;
+    end if;
+    -- No change in item's sensitive fields?
+    if (TG_OP = 'UPDATE'
+        AND (OLD.jsonb->'holdingsRecordId' IS NOT DISTINCT FROM NEW.jsonb->'holdingsRecordId')
```

Comment:
What does this check do? Does it mean that if an item moves holding this trigger won't be run?

Reply:
@marcjohnson-kint the `IS NOT DISTINCT FROM` instructions is the same as `=` but it also null safe. 

It means when we use `=` operator to compare `NULL` with something we will always get `NULL` as the operation result and if statement won't trigger.

But  `IS NOT DISTINCT FROM` allow us to compare NULLs safely, so `NULL IS NOT DISTINCT FROM 'Hello'` will return False vs `NULL` when we use `=`.

Reply:
If both the item's perm and temp locations are null before and after the update then the holding record determines the effective location. If the holdings record doesn't change then the effective location doesn't change and we can copy it from the old item record. This is what this if clause does.
If the item moves to a different holding then we need to query the new holding record and must skip this if clause.

Reply:
Why is this section protected by a check for if the holdings record ID has not changed?

Reply:
I think I'm failing to follow this logic.

Does this check say, only re-calculate the effective location if the holdings record ID has not changed and the item temporary and permanent location used to be null?

How does that match the intent expressed in the comment, that no relevant item properties have changed?

Reply:
@julianladisch 

> If both the item's perm and temp locations are null before and after the update

Where in this check does it check that the new item permanent or temporary location is null?

Reply:
In line 38 it checks whether the new item permanent or temporary location is null. If it isn't null it will return in line 40.

Reply:
The `if (TG_OP = 'UPDATE'` clause checks if there were no changes in the relevant item fields. If there were no changes then it uses the old effective location, no need to do any calculation.
Otherwise the execution continues after the if clause and runs a select query to fetch the holdings record and then calculates the effective location.